input/postgres/relation_stats_ext.go

```diff
@@ -0,0 +1,96 @@
+package postgres
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+
+	"github.com/lib/pq"
+	"github.com/pganalyze/collector/state"
+	"github.com/pganalyze/collector/util"
+)
+
+const extendedStatisticsSQLInheritedField = "null"
+const extendedStatisticsSQLpg15InheritedField = "se ->> 'inherited'"
+
+// pg_get_statisticsobjdef_expressions(s.oid) provides pg_stats_ext.exprs without accessing pg_stats_ext
+const extendedStatisticsSQL = `
+WITH relation_stats_ext AS (
+	SELECT se ->> 'statistics_schemaname' statistics_schemaname,
+		   se ->> 'statistics_name' statistics_name,
+		   %s inherited,
+		   se ->> 'n_distinct' n_distinct,
+		   se ->> 'dependencies' dependencies
+	  FROM %s
+)
+SELECT c.oid,
+	   n.nspname,
+	   s.stxname,
+	   (SELECT array_agg(k) FROM unnest(s.stxkeys) k) stxkeys,
+	   COALESCE(pg_get_statisticsobjdef_expressions(s.oid)::text[], ARRAY[]::text[]) exprs,
+	   s.stxkind,
+	   sd.inherited,
+	   sd.n_distinct,
+	   sd.dependencies
+  FROM pg_catalog.pg_statistic_ext s
+  JOIN pg_class c ON (s.stxrelid = c.oid)
+  JOIN pg_namespace n ON (s.stxnamespace = n.oid)
+  LEFT JOIN relation_stats_ext sd ON (sd.statistics_schemaname = n.nspname AND sd.statistics_name = s.stxname)
+ WHERE ($1 = '' OR (n.nspname || '.' || c.relname) !~* $1)
+`
+
+func GetRelationStatsExtended(ctx context.Context, logger *util.Logger, db *sql.DB, postgresVersion state.PostgresVersion, server *state.Server, globalCollectionOpts state.CollectionOpts, systemType string, dbName string) (state.PostgresRelationStatsExtendedMap, error) {
+	var sourceTable string
+	var inheritedField string
+
+	if postgresVersion.Numeric >= state.PostgresVersion15 {
+		inheritedField = extendedStatisticsSQLpg15InheritedField
+	} else {
+		inheritedField = extendedStatisticsSQLInheritedField
+	}
+
+	if StatsHelperExists(ctx, db, "get_relation_stats_ext") {
+		logger.PrintVerbose("Found pganalyze.get_relation_stats_ext() stats helper")
+		sourceTable = "pganalyze.get_relation_stats_ext() se"
+	} else {
+		if systemType != "heroku" && !connectedAsSuperUser(ctx, db, systemType) && globalCollectionOpts.TestRun {
+			logger.PrintInfo("Warning: Limited access to extended table statistics detected in database %s. Please set up"+
+				" the monitoring helper function pganalyze.get_relation_stats_ext (https://github.com/pganalyze/collector#setting-up-a-restricted-monitoring-user)"+
+				" or connect as superuser, to get extended statistics for all tables.", dbName)
+		}
+		sourceTable = "(SELECT row_to_json(pse.*)::jsonb se FROM pg_catalog.pg_stats_ext pse) _"
```

Comment:
[note for reviewer] As we're returning jsonb with the helper function, we also need to return jsonb when we directly use `pg_stats_ext` here. `row_to_json(pse.*)::jsonb se` will allow me to specify the jsonb object using `se->>'something'`.

Reply:
Another option is to also push that first CTE into the UDF, which would let you just use a simple table reference here:

```sql
WITH unwanted_pg_stats_ext_columns AS (
  SELECT
    array_agg(attname) AS cols
  FROM
    pg_attribute
  WHERE
    attrelid = 'pg_stats_ext'::regclass AND attname NOT IN ('statistics_schemaname','statistics_name','exprs','inherited','dependencies')            
),
pg_stats_ext_json AS (
  SELECT row_to_json(se.*)::jsonb - unwanted_pg_stats_ext_columns.cols
    FROM pg_catalog.pg_stats_ext se, unwanted_pg_stats_ext_columns
)
SELECT
  se ->> 'statistics_schemaname' AS statistics_schemaname,
  se ->> 'statistics_name' AS statistics_name,
  COALESCE(se ->> 'exprs', '[]') AS exprs,
  se ->> 'inherited' AS inherited,
  se ->> 'n_distinct' n_distinct,
  se ->> 'dependencies' dependencies
FROM
  pg_stats_ext_json
```

That also streamlines a bunch of the version-specific dynamic column handling in the CTE SQL (lines 13-16 in this file can be deleted and the stuff that depends on them can be simplified), so I think it's worth doing, but I don't feel super-strongly about it.

Note that I had to change `exprs` from your `ARRAY[]::text[]` to a plain `'[]'` since `se ->> 'exprs'` is a string in this context, but I think that still works.

I tested this locally on both 14 and 15 and it works fine.

Reply:
Ugh so tl;dr `COALESCE(se ->> 'exprs', '[]') AS exprs` won't work, not because you are wrong, but because I was wrong to start with; I only tested with the case of the column being `ARRAY[]::text[]` (aka exprs null), and now I test with the case with something in exprs, it started raising errors `"exprs": pq: unable to parse array; expected '{' at offset 0` in golang side, likely because it can't convert it to `pq.Array`.

Now I'm very much in the 🐇 🕳️ of converting `jsonb` array to the text array and just started hating the world. I kinda got something with LATERAL join but not sure if this is good, looks ugly. Also tried something with a subquery in the SELECT clause (by using something similar to [the function](https://stackoverflow.com/questions/45243186/how-to-cast-json-array-to-text-array)), but that also keeps giving me errors here and there with null case.
I'll sleep on this for now but if you have any good ideas, I'm all ears.

Reply:
Oh man, this is tricky. Starting from your LATERAL hint, I came up with:

```sql
WITH stats_ext AS (                                               
  SELECT
    row_number() over() AS row_id,
    statistics_schemaname,
    statistics_name,
    dependencies,
    n_distinct,
    most_common_val_nulls,
    most_common_freqs,
    most_common_base_freqs,
    row_to_json(se.*)::jsonb ->> 'inherited' AS inherited,
    row_to_json(se.*)::jsonb -> 'exprs' AS exprs
  FROM
    pg_stats_ext se
)
SELECT
  se1.statistics_schemaname,
  se1.statistics_name,
  se1.dependencies,
  se1.n_distinct,
  se1.most_common_val_nulls,
  se1.most_common_freqs,
  se1.most_common_base_freqs,
  se1.inherited,
  se_exprs.exprs
FROM stats_ext AS se1, LATERAL (
  SELECT
    coalesce(array_agg(exprs), array[]::text[]) AS exprs
  FROM (
    SELECT
      jsonb_array_elements_text(coalesce(se2.exprs, '[]')) AS exprs
    FROM
      stats_ext se2 WHERE se1.row_id = se2.row_id
  )
) AS se_exprs;
```

I had to add a `row_number()` window function to join these, since I'm not sure if there's a join key. I guess we could use (statistics_schemaname, statistics_name) instead? I guess that should work. This seems to do the right thing for me both when exprs is and is not available.

Reply:
Ah sorry I should have updated, but yeah this is going to be tricky but I'm actually going to use the Lukas's original way of using `pg_get_statisticsobjdef_expressions` function to obtain this instead of trying to deal with jsonb thingy. In this way, I don't need to get `exprs` from pg_stats_ext

Reply:
(Lukas kinda said it in https://github.com/pganalyze/collector/pull/476#discussion_r1454370154)

Reply:
Ah got it. I saw the comment but didn't put together that it meant we don't need to handle `exprs` like this.